File: Structural/Decorator/src/examplejava/decorators/DataSourceDecorator.java
package examplejava.decorators;

import examplejava.datasource.DataSource;

//Base Decorator
public class DataSourceDecorator implements DataSource {
    private DataSource wrappee;

    public DataSourceDecorator(DataSource wrappee) {
        this.wrappee = wrappee;
    }

    @Override
    public void writeData(String data) {
        wrappee.writeData(data);
    }

    @Override
    public String readData() {
        return wrappee.readData();
    }
}

File: Structural/Decorator/src/examplekotlin/decorators/DataSourceDecorator.kt
package examplekotlin.decorators

import examplejava.datasource.DataSource

open class DataSourceDecorator constructor(val wrappee: DataSource) : DataSource {

    override fun writeData(data: String?) {
        wrappee.writeData(data)
    }

    override fun readData(): String {
        return wrappee.readData()
    }
}